metric_learn/scml.py

```diff
@@ -0,0 +1,624 @@
+"""
+Sparse Compositional Metric Learning (SCML)
+"""
+
+from __future__ import print_function, absolute_import, division
+import numpy as np
+from .base_metric import _TripletsClassifierMixin, MahalanobisMixin
+from ._util import components_from_metric
+from sklearn.base import TransformerMixin
+from .constraints import Constraints
+from sklearn.preprocessing import normalize
+from sklearn.neighbors import NearestNeighbors
+from sklearn.cluster import KMeans
+from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
+from sklearn.utils import check_array, check_random_state
+import warnings
+
+# hack around lack of where in older numpy versions
+try:
+  np.sum([[0, 1], [1, 1]], where=[False, True], axis=1)
+except TypeError:
+  def sum_where(X, where):
+    return np.sum(X[where])
+else:
+  def sum_where(X, where):
+    return np.sum(X, where=where)
```

Comment:
Do you think this is worth it? You only use it to compute the objective function, and I don't think the complexity is wildly different (according to quick timing experiment, `np.sum(X[where])` is slightly slower for dense `where` but slightly faster in the sparse case). Considering that the where option was just introduced in the last version of numpy, I think that simply using `np.sum(X[where])` is good enough for now.

Reply:
ok, sounds good!